blue_bringup/scripts/gripper_calibration_service.py

```diff
@@ -0,0 +1,149 @@
+
+import rospy
+import numpy as np
+import actionlib
+from blue_msgs.srv import GripperPositionCalibration
+from controller_manager_msgs.srv import SwitchController
+from controller_manager_msgs.srv import LoadController
+from controller_manager_msgs.srv import ListControllers
+from std_srvs.srv import Trigger, TriggerRequest, TriggerResponse
+from blue_msgs.msg import MotorState
+from std_msgs.msg import Float64MultiArray
+from control_msgs.msg import (
+    GripperCommandAction,
+    GripperCommandGoal,
+)
+
+def get_gripper_state(motor_names):
+    motor_state_msg = rospy.wait_for_message("blue_hardware/motor_states", MotorState)
+    for i, n in enumerate(motor_state_msg.name):
+        if n == motor_names[-1]:
+            return motor_state_msg.position[i]
+
+def handle_load_controllers():
+    unload_list = []
+    load_controllers_list = ['blue_controllers/gripper_torque_controller', 'blue_controllers/gripper_controller']
+    loaded_controllers    = list_controllers()
+    rospy.logerr(loaded_controllers.controller)
+    for controller in load_controllers_list:
+        load = True
+        for loaded in loaded_controllers.controller:
+            rospy.logerr(loaded.name)
+            rospy.logerr(controller)
+            rospy.logerr(controller)
+
+            if loaded.name == controller:
+                # controller already loaded
+                if loaded.state == "running":
+                    return False
+                load = False
+                break
+        if load:
+            # controller not loaded yet
+            controller_response = load_controller(controller)
+            unload_list.append(controller)
+    return unload_list
+
+
+def handle_calibration_service(input_request):
+    # Calibrate joints with startup angles
+    rospy.loginfo("Starting calibration...")
+    try:
+        unload_list = handle_load_controllers()
+        if unload_list == False:
+            return TriggerResponse(False, "Gripper Calibration Failed. Gripper resource must be free.")
+        gripper_controller_response = switch_controller(['blue_controllers/gripper_torque_controller'], [], 2)
+
+        # Close the gripper
+        cmd = Float64MultiArray()
+        cmd.data = [5.0]
+        for i in range(5):
+            cmd_pub.publish(cmd)
+            rospy.sleep(0.1)
+
+        # Wait until gripper is closed/stops moving
+        gripper_state = get_gripper_state(motor_names)
+        current_position = gripper_state + 10
+        while not rospy.is_shutdown() and np.abs(current_position - gripper_state) > 1e-5:
+            cmd_pub.publish(cmd)
+            current_position = gripper_state
+            gripper_state = get_gripper_state(motor_names)
+
+        # Stop applying force
+        cmd.data = [0.0]
+        for i in range(5):
+            cmd_pub.publish(cmd)
+            rospy.sleep(0.1)
+
+        # Calibrate gripper to closed position
+        # TODO: don't hardcode this value -- read from yaml? URDF?
```

Comment:
should we implement this TODO with this PR?

i think i added it and can probably do it if you want oops

Reply:
I looked into trying to do this and couldn't find a clean way to do it. Do you want to give it a try?

Reply:
We can just shelve this for a future PR